# Variables and values

Elements of Data Science

by [Allen Downey](https://allendowney.com)

[MIT License](https://opensource.org/licenses/MIT)

## Numbers

This notebook introduces the most fundamental tools for working with data: representing numbers and other values, and performing arithmetic operations.

Python provides tools for working with numbers, strings, dates, times, and locations (in latitude and longitude).

Let's start with numbers.  Python can handle several types of numbers, but the two most common are:

* `int`, which represents integral values like `3`, and
* `float`, which represents numbers that have a fraction part, like `3.14159`.

Most often, we use `int` to represent counts and `float` to represent measurements.

In [1]:
3

3

In [2]:
3.14159

3.14159

## Arithmetic

The operators that perform addition and subtraction are `+` and `-`:

In [3]:
2 + 1

3

In [4]:
2 - 1

1

The operators that perform multiplication and division are `*` and `/`:

In [5]:
2 * 3

6

In [6]:
2 / 3

0.6666666666666666

And the operator for exponentiation is `**`:

In [7]:
2**3

8

These operators follow the rules of precedence you might have learned as "PEMDAS":

* Parentheses before
* Exponentiation before
* Multiplication and division before
* Addition and subtraction

So in this expression:

In [8]:
1 + 2 * 6

13

The multiplication happens first.  If that's not what you want, you can use parentheses to make the order of operations explicit:

In [9]:
(1 + 2) * 6

18

## Math functions

Python provides functions that compute all the usual mathematical functions, like `sin` and `cos`, `exp` and `log`.

Actually, it provides two versions of these functions, in two different libraries, called `math` and `numpy`.  The version we'll use is `numpy`, which stands for "numerical python", and is pronounced "num' pie".

Before we can use `numpy`, we have to "import" it like this:

In [10]:
import numpy as np

`numpy` provides `pi`, which is a floating-point approximation of the mathematical constant $\pi$:

In [11]:
np.pi

3.141592653589793

So, for example, $\cos(\pi) = -1$.

In [12]:
np.cos(np.pi)

-1.0

But in general, floating-point numbers are only approximate.  The approximation of $\pi$ has about 16 digits.

If we compute $sin(\pi)$, we expect the result to be zero.

In [13]:
np.sin(np.pi)

1.2246467991473532e-16

The result is in scientific notation; the `e` in the middle stands for "exponent".  So the result is about $1.22 \times 10^{-16}$, which is very small, but not exactly 0.

numpy also provides `log`, which computes the natural logarithm, and `exp`, which raises the constant `e` to a power.

In [14]:
np.exp(1)

2.718281828459045

In [15]:
np.log(100)

4.605170185988092

## Variables

The following statement assigns the `int` value 5 to a variable named `x`:


In [16]:
x = 5

A variable is a name that refers to a value.  The variable we just created has the name `x` and the value 5.

If we use `x` as part of an arithmetic operation, it represents the value 5:

In [17]:
x + 1

6

In [18]:
x**2

25

We can also use `x` with `numpy` functions:

In [19]:
np.exp(5)

148.4131591025766

As an example, suppose we have the following formula for computing compound interest [from Wikipedia](https://en.wikipedia.org/wiki/Compound_interest#Periodic_compounding):

"The total accumulated value, including the principal sum $P$ plus compounded interest $I$, is given by the formula:

$V=P\left(1+{\frac {r}{n}}\right)^{nt}$

where:

* $P$ is the original principal sum
* $V$ is the total accumulated value
* $r$ is the nominal annual interest rate
* $n$ is the compounding frequency
* $t$ is the overall length of time the interest is applied (expressed using the same time units as $r$, usually years).

"Suppose a principal amount of \$1,500 is deposited in a bank paying an annual interest rate of 4.3\%, compounded quarterly.
Then the balance after 6 years is found by using the formula above, with

In [20]:
P = 1500
r = 0.043
n = 4
t = 6

We can compute the total accumulated value by translating the mathematical formula into Python syntax:

In [21]:
P * (1 + r/n)**(n*t)

1938.8368221341054

**Exercise:** Continuing the example from Wikipedia:

"Suppose the same amount of \$1,500 is compounded biennially", so `n = 1/2`.  

What would the total value be after 6 years?

In [22]:
# Solution

n = 1/2
P * (1 + r/n)**(n*t)

1921.2360840000006

**Exercise:** If interest is compounded continuously, the value after time $t$ is given by the formula:

$V=P~e^{rt}$

Translate this function into Python and use it compute the value of the investment in the previous example with continuous compounding.

In [23]:
# Solution

P * np.exp(r*t)

1941.5082279363567

**Exercise** Applying your algebra skills, solve the previous equation for $r$.  Now use the formula you just derived to answer this question.

Harvard's tuition in 1970 was \$4,070 (not including room, board, and fees).  

In 2019 it is \$46,340.  What was the annual rate of increase over that period?  (Treating it as if it had compounded continuously.)

In [24]:
# Solution

P = 4070
V = 46340
t = (2019 - 1970)
np.log(V/P) / t

0.049640051419382165